crates/cranelift/src/bounds_checks.rs

```diff
@@ -31,26 +33,155 @@ use cranelift_frontend::FunctionBuilder;
 use wasmtime_environ::Unsigned;
 use Reachability::*;
 
+/// The kind of bounds check to perform when accessing a Wasm linear memory or
+/// GC heap.
+///
+/// Prefer `BoundsCheck::*WholeObject` over `BoundsCheck::Field` when possible,
+/// as that approach allows the mid-end to deduplicate bounds checks across
+/// multiple accesses to the same GC object.
+#[derive(Debug)]
+pub enum BoundsCheck {
+    /// Check that this one access in particular is in bounds:
+    ///
+    /// ```ignore
+    /// index + offset + access_size <= gc_heap_bound
```

Comment:
Mind renaming `s/gc_heap_bound/bound/`? (to emphasize this isn't just GC accesses)